template/node12/index.js

```diff
@@ -30,62 +30,127 @@ app.use(function addDefaultContentType(req, res, next) {
+            listenerHandlers = manifest.listeners || {};
+            manifest.widgets = Object.keys(widgetHandlers);
+            manifest.listeners = Object.keys(listenerHandlers);
+
+            res.status(200).json({ manifest: manifest, stats: { ui: Number(uiStopTime - uiStartTime) } });
         })
-        .then(newUi => {
-            uiStopTime = process.hrtime.bigint();
-            res.status(200).json({ data: newData, ui: newUi, stats: { listeners: Number(listenersStopTime - listenersStartTime), ui: Number(uiStopTime - uiStartTime) } });
+        .catch(err => {
+            const err_string = err.toString ? err.toString() : err;
+            console.error("handleAppManifest:", err_string);
+            res.status(500).send(err_string);
+        });
+}
+
+async function handleAppWidget(req, res) {
+
+
+    let { widget, data, props } = req.body;
+
+    let uiStartTime = process.hrtime.bigint();
+
+    let possibleFutureRes = widgetHandlers[widget](data, props);
+
+    return Promise.resolve(possibleFutureRes)
+        .then(widget => {
+            let uiStopTime = process.hrtime.bigint();
+
+            res.status(200).json({ widget: widget, stats: { ui: Number(uiStopTime - uiStartTime) } });
         })
         .catch(err => {
-            res.status(500).send(err.toString ? err.toString() : err);
+            const err_string = err.toString ? err.toString() : err;
+            console.error('handleAppWidget:', err_string);
+            res.status(500).send(err_string);
         });
 }
 
-app.post('/*', middleware);
+/**
+ * This is the main entry point for the OpenFaaS function.
+ *
+ * This function will call the index.js file of the application
+ * when the page change.
+ * If an event is triggered, the matched event function provided by the app is triggered.
+ * The event can be a listener or a widget update.
+ */
+async function handleAppListener(req, res) {
+
+    let newData = {};
+
+    let { action, data, props, event } = req.body;
+
+    let listenersStartTime = process.hrtime.bigint();
+
+    /*
+        listeners file need to exactly math with action name
+    */
+    if (Object.keys(listenerHandlers).includes(action)) {
+        let possibleFutureRes = listenerHandlers[action](data, props, event);
+
+        return Promise.resolve(possibleFutureRes)
+            .then(data => {
+                let listenersStopTime = process.hrtime.bigint();
+                newData = data;
+
+                res.status(200).json({ data: newData, stats: { listeners: Number(listenersStopTime - listenersStartTime) } });
+            })
+            .catch(err => {
+                const err_string = err.toString ? err.toString() : err;
+                console.error('handleAppAction:', err_string);
+                res.status(500).send(err_string);
+            });
+    } else {
+        console.error(`No listener found for action ${action} in app manifest.`);
+    }
+}
+
+//middleware to catch ressource
+app.post('/', middleware);
```

Comment:
```suggestion
app.post('/*', middleware);
```

If I properly remember the work on the resources, the star "*" is important so that the middleware catches every resource under every path.
Tell me if I am wrong

Reply:
@shiipou